zilliqa/src/api/zil.rs

```diff
@@ -1242,9 +1191,77 @@ fn get_sharding_structure(_params: Params, _node: &Arc<Mutex<Node>>) -> Result<(
     todo!("API getshardingstructure is not implemented yet");
 }
 
+fn get_smart_contract_state_internal(
+    address: Address,
+    requested_varname: Option<String>,
+    requested_indices: Option<HashSet<String>>,
+    node: &Arc<Mutex<Node>>,
+) -> Result<Value> {
+    let node = node.lock().unwrap();
+
+    // First get the account and check that its a scilla account
+    let block = node
+        .get_block(BlockId::latest())?
+        .ok_or_else(|| anyhow!("Unable to get latest block!"))?;
+
+    let state = node.get_state(&block)?;
+    let account = state.get_account(address)?;
+
+    let result = json!({
+        "_balance": ZilAmount::from_amount(account.balance).to_string(),
+    });
+    let Value::Object(mut result) = result else {
+        unreachable!()
+    };
+
+    let is_scilla = account.code.scilla_code_and_init_data().is_some();
+    if is_scilla {
+        let limit = node.config.state_rpc_limit;
+
+        let trie = state.get_account_trie(address)?;
+        for (i, (k, v)) in trie.iter().enumerate() {
+            if i >= limit {
+                return Err(anyhow!(
+                    "State of contract returned has size greater than the allowed maximum"
+                ));
+            }
+
+            let (var_name, indices) = split_storage_key(&k)?;
+            if let Some(ref x) = requested_varname {
+                if *x != var_name {
+                    continue;
+                }
+            }
+            let mut var = result.entry(var_name.clone());
+
+            for index in indices {
```

Comment:
The indices are ordered and easily seekable. We could save a lot of work by only querying the indices we need from the state.

As it stands, `GetSmartContractSubState` requires almost the same amount of work for us as `GetSmartContractState`.